streaming/src/test/scala/org/apache/spark/streaming/scheduler/RateControllerSuite.scala

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.streaming.scheduler
+
+import scala.collection.mutable
+import scala.reflect.ClassTag
+import scala.util.control.NonFatal
+
+import org.scalatest.Matchers._
+import org.scalatest.concurrent.Eventually._
+import org.scalatest.time.SpanSugar._
+
+import org.apache.spark.streaming._
+import org.apache.spark.streaming.scheduler.rate.RateEstimator
+
+class RateControllerSuite extends TestSuiteBase {
+
+  override def actuallyWait: Boolean = true
+
+  test("rate controller publishes updates") {
+    val ssc = new StreamingContext(conf, batchDuration)
+    val dstream = new MockRateLimitDStream(ssc, Seq(Seq(1)), 1)
+    val output = new TestOutputStreamWithPartitions(dstream)
```

Comment:
This old style of using `TestSuiteBase` and defining `streamingContext` in every test and making runStreams stop the context is actually very flaky, as failure before `runStream` will end up leaking an active SparkContext and failing all subsequent Spark tests. Its being used in many places and I hope to upgrade them at some point. So the two alternatives of using TestSuiteBase are

1. Use [`usingStreamingContext`](https://github.com/apache/spark/blob/1f6b0b1234cc03aa2e07aea7fec2de7563885238/streaming/src/test/scala/org/apache/spark/streaming/TestSuiteBase.scala#L270) - This ensure the context are stopped no matter what fails. But it ends up creating a new SpakrContext for every test. 

2. The 2nd alternative is to share SparkContext and use before-after methods as I have done it in this [suite](https://github.com/apache/spark/blob/master/extras/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisStreamSuite.scala). 

Please update the testsuites here with either of the two methods.